text/1011-query-rewrite.rst

```diff
@@ -0,0 +1,322 @@
+such as audit logs or to automatically generate external change events.
+
+
+Requirements
+============
+
+Although there is no direct dependency, this RFC requires RFC 1010 (globals)
+to be implemented to be practical.  See `Interaction with globals`_ below
+for details.
+
+
+Specification
+=============
+
+This RFC proposes to add a new schema item type -- *Policy* -- that is
+defined in the context of an object type and specifies the rules of rewriting
+queries that refer to the enclosing object type.
+
+CREATE POLICY
+-------------
+
+Define a new query rewrite policy for a given object type.
+
+Required capabilities: DDL.
+
+Synopsis::
+
+    {CREATE|ALTER} OBJECT TYPE <type-name> "{"
+        CREATE POLICY
+            ON { SELECT | INSERT | UPDATE | DELETE } [, ... ]
+            [ WHEN <condition> ]
+            <action>
+    "}"
+
+    # where <action> can be one of:
+
+    PERMIT <element-expr>
+
+    RESTRICT TO <element-expr>
+
+    CHECK <element-expr>
+
+    DO AFTER <set-expr>
+
+    DO ON COMMIT <set-expr>
+
+Policies apply to the given list of statement types.  Not all actions are
+valid for all statement types.  See table below for legal combinations.
+
+The optional ``<condition>`` expression is evaluated for every object
+affected by the statement and the policy is applied only if the expression
+evaluates to a set containing at least one ``true`` value.
+
+``<element-expr>`` is evaluated for every object affected by the statement
+the same way as a ``FILTER`` clause is evaluated, and can use abbreviated
+paths, for example::
+
+    POLICY ON SELECT RESTRICT TO .rating != 'R'
+
+``<set-expr>`` is evaluated for the entire set of objects affected by the
+statement and can refer to the set as ``__result__``, for example::
+
+    POLICY ON UPDATE DO AFTER (
+        FOR obj IN __result__ UNION (
+        INSERT AuditLog {
+            text := 'updated ' ++ obj.name
+        })
+    )
+
+    # or
+
+    POLICY ON INSERT DO AFTER (
+        INSERT EventLog {
+            text := 'inserted ' ++ count(__result__) ++ ' things'
+        }
+    )
+
+The ``PERMIT`` and ``RESTRICT TO`` actions are *filtering* actions and define
+which elements of the object set are visible in a query.  If at least one
+filtering policy is defined for a type, then all elements become invisible by
+default and must be made visible by at least one ``PERMIT`` policy.
+
+The filtering actions are defined as follows:
+
+- ``PERMIT`` -- add all objects, for which the expression evaluates to a
+  set containing at least one ``true`` value, to the set of visible objects.
+  ``PERMIT`` policies are combined using the ``OR`` operator.
+
+- ``RESTRICT TO`` -- restrict the set of visible objects to only objects for
+  which the expression evaluates to a set containing at least one ``true``
+  value.  ``RESTRICT`` policies are combined using the ``AND`` operator and
+  are applied *after* all ``PERMIT`` policies.
```

Comment:
> `policy on select when (exists global user_id) permit (.owner.id = global user_id)`

As per [this comment](https://github.com/edgedb/rfcs/pull/49#issuecomment-1017654332) `exists global user_id` will read as `exists assert_exists(global user_id)`. Will we make an exception when using `exists` or `when`, or am I misunderstanding how it works?

> `policy on select when (.private) restrict to (global user = .owner);`

This will also need something like:
```
policy on select when (.private OR global system_admin) restrict to (global user = .owner);
```
Which is annoying. I think [this proposal](https://github.com/edgedb/rfcs/pull/50#issuecomment-1018486921) could be better than `RESTRICT TO` here.

Reply:
> This will also need something like:
> `global system_admin`

Did you mean `not global system_admin`?

Not really.  The idea is that certain roles will be given permission to bypass filters altogether.

Reply:
> > This will also need something like:
> > `global system_admin`
> 
> Did you mean `not global system_admin`?
> 

Ah, yes. Like I've said interactions between WHEN/RESTRICT/PERMIT are error-prone an confusing.

> Not really. The idea is that certain roles will be given permission to bypass filters altogether.

I don't see "bypassing roles" used in applications. It's more like for dump/restore or disaster recovery.

Imagine an `analytic` and `moderator` roles which need to overcome this specific restriction but not all of them.